README.md
# Y PPA Manager
## Y PPA Manager features:
- Add PPA
- Remove PPA
- Purge PPA
- Search for packages in Launchpad PPAs: a regular search which is faster, but doesn't display exact package matches and comes with less details and a deep search which displays exact package matches. In the search, you can also see if a PPA is already added on your system or not and if a package is already installed (and the installed version). You can perform the following actions on a PPA listed in the search results: add it, list packages in the PPA, open PPA in browser, download packages, install a package using the built-in installer (if the PPA is not already added, it will be added)
- Update single PPAs - without running a full "apt-get update", which should be a lot faster (and especially useful for computers with slow Internet connections)
- List packages in PPAs enabled on your computer
- Edit PPA source file
- Remove duplicate PPAs
- Import all missing GPG keys
- Fix GPG BADSIG errors
- Backup an restore PPAs (automatically imports missing GPG keys)
- Re-enable working PPAs after Ubuntu upgrade: when you upgrade to a newer Ubuntu release, the PPAs are disabled so using this feature, the PPAs that work with the new Ubuntu version you're using are re-enabled, leaving the others disabled
- Update release name in working PPAs: somewhat similar to the feature above, this one is useful if you've backup up the PPAs in say Ubuntu Precise and restored them in Ubuntu Quantal (just an example) - in this case, using this feature you can replace "precise" in each PPA source with "quantal", but only for the PPAs that have packages for Quantal.
- Desktop integration: notifications, Unity quicklists, indicator and HUD support

### Settings:
- PPA Purge behavior: auto - don't require any user input; manual - opens a terminal window asking the user how to solve the issue (this is the default and highly recommended behavior).
- Ubuntu version: this only affects the search. So if you want the Y PPA Manager search to display packages for some other Ubuntu version, simply change the Ubuntu version here. Supported versions: karmic, lucid, maverick or natty (use the Ubuntu version names for Linux Mint too!).


CHANGES
0.9.9.3:
fix bug with dpkg being reported as running in some cases when it wasn't actually running, this causing some Y PPA Manager features not to work
fix bug in upgrade dialog with GTK2
a few other small bug fixes

0.9.9.2:
- Ubuntu 14.04 Trusty Tahr and Linux Mint 16 Petra support
- Added install button on Manage PPAs -> List Packages

0.9.9.1:
- Ubuntu 13.10: fix policy kit issue with GNOME 3.10
- Ubuntu 13.10: fix main window issues with GTK 3.8.3+
- Ubuntu 13.10: fix appindicator
- Notification fixes for GNOME Shell
- Fix backup not backing up sources.list PPAs
- use the deep search as default and add checkbox for the basic search instead
- use single click to open items in the main dialog
- added Faience indicator icon (thanks to Giovani Cascaes!)
- Fixed: Re-enable working PPAs after Ubuntu upgrade to work on non-english systems (lp: #1244986)
- Fixed: a couple more small bugs

0.9.9:
- support for Linux Mint 15 Olivia as well as the latest Ubuntu 13.10 Saucy Salamander

0.0.9.8:
- display notification when the upgrade and update release name PPA options are started
- add policykit-1 dependency - fixes LP: #1125240
- better way to determine currently logged in username and home dir
- fix backup path defaulting to root homedir instead of user homedir in the process
- sometimes y-pkexec may be called as root by y-ppa-manager so let's fix ~/.Xauthoriy for KDE when that happens
- lots of notification fixes: fix search notification not being displayed when the search item contains spaces, display notification before launching the Manage PPAs dialog because it may take a while before it s displayed, fix notifications being displayed before ppa-upgrade ran (update release name in working ppas, etc), more
- show advanded dialog if user cancels duplicate removal

0.0.9.7:
- copy the PPA link to the clipboard instead of opening the PPA in the default web browser. this change is required because since we've started using pkexec, opening links in the browser doesn't work with Chromium, Google Chrome or Opera. It also doesn't work with Firefox if Firefox isn't running. this is a work-around for bug lp: #1079744

0.0.9.6:
- elementary OS Luna support

0.0.9.5:
- add keywords to the desktop file, fixes LP: #1077793
- add Linux Mint 14 "Nadia" support

0.0.9.4:
- temporary work-around for checking display for the loggeduser variable, since the display can be e.g. :5 or :5.0, etc. and also avoid grep fiding the time instead of the display by using ($DISPLAY) instead of $DISPLAY. This affects the notifications and opening a PPA in the default web browser (which will now hopefully work for all users).
- Don't use an extra file for helpcheck (thanks Satya)
- Fix advanced PPA search not working properly with PPAs which have a dot in the name
- other fixes

0.0.9.3:
- minor bug fixes, including: xdg-open for ppa links shouldn't open browser as root, ppa add check fixes
- add work-around for the manual ppa add dialog bug: when add-apt-repository is used with -y, it adds the PPA even if it doesnt exist
- md5sums support for update-ppa
- add regex search for the YAD list dialogs

0.0.9.2:
- fixed a bug with search results dialog and apt-get
- implement transient notifications for GNOME Shell
- fix issue with PPA Restore function not working with spaces
- fixed bugs occuring for PPAs with "." or "+" characters in their names
- update-ppa: add support for updating PPAs having older release names
- update-ppa: fix bug with update-ppa not working on Linux Mint
- update-ppa: support for more repository formats (experimental). Examples: update-ppa http://archive.ubuntu.com/ubuntu or update-ppa http://archive.ubuntu.com/ubuntu quantal main
- update-ppa: Support specifying custom codename and section for a ppa. Example: update-ppa ppa:user/repository quantal main
- update-ppa: Download compressed .gz files instead of normal files
- update-ppa: Support force update lists without adding the repository
- merge "PPA Remove", "List Packages" and "PPA Purge" into a new Manage PPAs dialog
- Manage PPAs: initial support for Launchpad PPAs (only Launchpad!) from /etc/apt/sources.list (experimental). Note: the duplicate PPAs removal, "Re-enable working PPAs after Ubuntu upgrade" and "Update release name in working PPAs" features don't yet support this.
- Manage PPAs: option to view/edit the .list source file
- Manage PPAs: option to update a single PPA
- list PPAs: display the packages for the Ubuntu version available in the .list file, not from the Ubuntu version the user is currently running (it might be different - e.g.: a package is not available in PPA X for Ubuntu Quantal, so an user running Ubuntu Quantal modifies the PPA source file to Ubuntu Precise)
- Manage PPAs: display the actual PPA name (e.g.: ppa:webupd8team/gthumb) in this dialog, instead of the PPA .list file name
- use policykit (pkexec) instead of gksu / kdesu for Ubuntu versions newer than 11.04 (11.04 not included)
- install_package.py should handle errors better


0.0.9.1:
- minor bug fixes

0.0.9.0:
- new "Install packages" item in the main UI that lets you quickly install packages (if you already know the package names). Separate multiple packages with a space. This python script was contributed by Lorenzo from atareao.es
- when you search for packages in Launchpad PPAs, you can select to install some packages directly from the search results dialog - if the PPA isn't already added on your system, it will be added and then the selected package(s) will be installed using the new "Install packages" python script mentioned above, that comes with a GUI, so the user can see if any errors occur.
- a new "PPA status" column in the basic search and advanced search results dialog that lets you know if the PPA is already added on your system or not
- Search results dialog: when searching for a package (in deep search only), Y PPA manager will display its version if it's installed, or it will let you know that the package is not installed.

install_package.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
#
# Copyright (C) 2012 Lorenzo Carbonell
# lorenzo.carbonell.cerezo@gmail.com
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
#
#

import pygtk
pygtk.require("2.0")
import gtk
import apt
import apt.progress.gtk2
import apt_pkg
from apt.cache import Cache
from apt.cache import LockFailedException

class InstallPackageDialog(gtk.Dialog):
	def __init__(self,packages = None):
		title = 'Y PPA Manager - Install package(s)'
		gtk.Dialog.__init__(self,title,None,gtk.DIALOG_MODAL | gtk.DIALOG_DESTROY_WITH_PARENT,(gtk.STOCK_CLOSE,gtk.RESPONSE_ACCEPT))
		self.set_wmclass = 'Y-PPA-Manager'
		self.set_icon_name('y-ppa-manager')
		self.set_position(gtk.WIN_POS_CENTER_ALWAYS)
		self.set_size_request(400, 200)
		self.set_resizable(False)
		self.connect('destroy', self.close_application)
		#
		vbox0 = gtk.VBox(spacing = 5)
		vbox0.set_border_width(5)
		self.get_content_area().add(vbox0)
		#
		frame2 = gtk.Frame()
		vbox0.add(frame2)
		table2 = gtk.Table(rows = 2, columns = 2, homogeneous = False)
		table2.set_border_width(5)
		table2.set_col_spacings(5)
		table2.set_row_spacings(5)
		frame2.add(table2)
		#
		label11 = gtk.Label('Package(s) to install:')
		label11.set_alignment(0,.5)
		table2.attach(label11,0,1,0,1, xoptions = gtk.EXPAND|gtk.FILL, yoptions = gtk.SHRINK)
		#
		#
		self.entry12 = gtk.Entry()
		if packages:
			self.entry12.set_text(packages)
		self.entry12.connect('key-press-event',self.on_key_press)
		table2.attach(self.entry12,1,2,0,1, xoptions = gtk.EXPAND|gtk.FILL, yoptions = gtk.SHRINK)
		#
		#
		button22 = gtk.Button('Install')
		button22.connect('clicked',self.install_package)
		table2.attach(button22,0,2,1,2, xoptions = gtk.EXPAND|gtk.FILL, yoptions = gtk.SHRINK)
		#
		self.progress = apt.progress.gtk2.GtkAptProgress()
		self.progress._expander.connect('activate',self.on_progress_activate)
		table2.attach(self.progress,0,2,2,3, xoptions = gtk.EXPAND|gtk.FILL, yoptions = gtk.SHRINK)
		#
		self.show_all()
		if packages:
			self.install_package(None)

	def on_key_press(self,widget,event):
		if self.entry12.get_text() != '' and (event.keyval == 65293 or event.keyval ==65421):
			self.install_package(None)
			
	def on_progress_activate(self,widget):
		if self.progress._expander.get_expanded():
			self.set_size_request(400, 200)
		else:
			self.set_size_request(800, 650)
		
	def close_application(self,widget):
		self.hide()
		self.destroy()
		
	def install_package(self,widget):
		cache = apt.cache.Cache(self.progress.open)
		packages = self.entry12.get_text().strip().split()
		for p in packages:
			print p
		inpackages = []
		nopackages = []
		noupgradables = []

		for package in packages:
			try:
				print "%s is installed %s and is upgradable %s"%(package,cache[package].is_installed,cache[package].is_upgradable)
				if cache[package].is_installed and not cache[package].is_upgradable:
					noupgradables.append(package)
				else:
					inpackages.append(package)
			except KeyError,e:
				print e
				nopackages.append(package)
		if len(noupgradables)>0 or len(nopackages)>0:
			message = "<b>Some errors found:</b>\n"
			if len(noupgradables)>0:
				if len(noupgradables)>1:
					tpackages = ', '.join(noupgradables)
					message += "\nThe packages: '%s'\n <b>can't be upgraded</b> or the latest version is already installed"%tpackages
				else:
					message += "\nThe package: %s <b>can't be upgraded</b> or the latest version is already installed"%noupgradables[0]
			if len(nopackages)>0:
				if len(nopackages)>1:
					tpackages = ', '.join(nopackages)
					message += "\nThere are <b>no packages</b> called:\n '%s'"%tpackages
				else:
					message += '\nThere is <b>no package</b> called "%s"'%nopackages[0]
				dbuttons=gtk.BUTTONS_OK
			if len(inpackages)>0:
				if len(inpackages)>1:
					tpackages = ', '.join(inpackages)
					message += "\n<b>Only</b> these packages: '%s'\n will be installed"%tpackages

				else:
					message += "\n<b>Only</b>: %s will be installed"%inpackages[0]
				message +="\n\n<b>Continue?</b>"
				ans_exit = False
				dbuttons=gtk.BUTTONS_YES_NO
			else:
				message += "\n\nThere is <b>no package</b> to install"
				message += "\nInstallation will <b>not continue</b>"
				ans_exit = True
				dbuttons=gtk.BUTTONS_OK
			md = gtk.MessageDialog(parent=self,
			flags=gtk.DIALOG_MODAL | gtk.DIALOG_DESTROY_WITH_PARENT,
			type=gtk.MESSAGE_ERROR,
			buttons=dbuttons,
			message_format=None)
			md.set_markup(message)
			ans = md.run()
			print ans
			if ans == -9 or ans_exit:
				md.destroy()
				exit(0)
			md.destroy()
		print inpackages
		if len(inpackages)>0:
			for inpackage in inpackages:
				if cache[inpackage].is_upgradable:
					cache[inpackage].mark_upgrade()
				else:
					cache[inpackage].mark_install(auto_fix=True, auto_inst=True, from_user=True)
			try:
				self.set_size_request(800, 650)
				self.progress.show_terminal(expanded=True)
				cache.commit(self.progress.fetch, self.progress.install)					
			except LockFailedException,e:
				print e
				md = gtk.MessageDialog(parent=self,
				flags= gtk.DIALOG_MODAL | gtk.DIALOG_DESTROY_WITH_PARENT,
				type=gtk.MESSAGE_ERROR,
				buttons= gtk.BUTTONS_OK,
				message_format="You didn't run this program as root or another package\nmanager such as Synaptic or apt-get is running")
				md.run()
				md.destroy()
				self.progress.show_terminal(expanded=False)
				self.set_size_request(400, 200)
		else:
			md = gtk.MessageDialog(parent=self,
			flags=gtk.DIALOG_MODAL | gtk.DIALOG_DESTROY_WITH_PARENT,
			type=gtk.MESSAGE_ERROR,
			buttons=gtk.BUTTONS_OK,
			message_format="You didn't enter any package to install!")
			md.run()
			md.destroy()
		self.progress.show_terminal(expanded=False)
		self.set_size_request(400, 200)

if __name__ == '__main__':
	import sys
	if len(sys.argv)>2:
		packages = (' '.join(sys.argv[1:])).strip()
	elif len(sys.argv) >= 2:
		packages = sys.argv[1]
	else:
		packages = None
	ipd = InstallPackageDialog(packages)
	ipd.run()


install_package-lucid.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
#
# Copyright (C) 2012 Lorenzo Carbonell
# lorenzo.carbonell.cerezo@gmail.com
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
#
#

import pygtk
pygtk.require("2.0")
import gtk
import apt
import apt.progress.gtk2
import apt_pkg
from apt.cache import Cache
from apt.cache import LockFailedException

class InstallPackageDialog(gtk.Dialog):
	def __init__(self,packages = None):
		title = 'Y PPA Manager - Install package(s)'
		gtk.Dialog.__init__(self,title,None,gtk.DIALOG_MODAL | gtk.DIALOG_DESTROY_WITH_PARENT,(gtk.STOCK_CLOSE,gtk.RESPONSE_ACCEPT))
		self.set_wmclass = 'Y-PPA-Manager'
		self.set_icon_name('y-ppa-manager')
		self.set_position(gtk.WIN_POS_CENTER_ALWAYS)
		self.set_size_request(400, 200)
		self.set_resizable(False)
		self.connect('destroy', self.close_application)
		#
		vbox0 = gtk.VBox(spacing = 5)
		vbox0.set_border_width(5)
		self.get_content_area().add(vbox0)
		#
		frame2 = gtk.Frame()
		vbox0.add(frame2)
		table2 = gtk.Table(rows = 2, columns = 2, homogeneous = False)
		table2.set_border_width(5)
		table2.set_col_spacings(5)
		table2.set_row_spacings(5)
		frame2.add(table2)
		#
		label11 = gtk.Label('Package(s) to install:')
		label11.set_alignment(0,.5)
		table2.attach(label11,0,1,0,1, xoptions = gtk.EXPAND|gtk.FILL, yoptions = gtk.SHRINK)
		#
		#
		self.entry12 = gtk.Entry()
		if packages:
			self.entry12.set_text(packages)
		self.entry12.connect('key-press-event',self.on_key_press)
		table2.attach(self.entry12,1,2,0,1, xoptions = gtk.EXPAND|gtk.FILL, yoptions = gtk.SHRINK)
		#
		#
		button22 = gtk.Button('Install')
		button22.connect('clicked',self.install_package)
		table2.attach(button22,0,2,1,2, xoptions = gtk.EXPAND|gtk.FILL, yoptions = gtk.SHRINK)
		#
		self.progress = apt.progress.gtk2.GtkAptProgress()
		self.progress._expander.connect('activate',self.on_progress_activate)
		table2.attach(self.progress,0,2,2,3, xoptions = gtk.EXPAND|gtk.FILL, yoptions = gtk.SHRINK)
		#
		self.show_all()
		if packages:
			self.install_package(None)

	def on_key_press(self,widget,event):
		if self.entry12.get_text() != '' and (event.keyval == 65293 or event.keyval ==65421):
			self.install_package(None)
			
	def on_progress_activate(self,widget):
		if self.progress._expander.get_expanded():
			self.set_size_request(400, 200)
		else:
			self.set_size_request(800, 650)
		
	def close_application(self,widget):
		self.hide()
		self.destroy()
		
	def install_package(self,widget):
		cache = apt.cache.Cache(self.progress.open)
		packages = self.entry12.get_text().strip().split()
		for p in packages:
			print p
		inpackages = []
		nopackages = []
		noupgradables = []

		for package in packages:
			try:
				print "%s is installed %s and is upgradable %s"%(package,cache[package].is_installed,cache[package].is_upgradable)
				if cache[package].is_installed and not cache[package].is_upgradable:
					noupgradables.append(package)
				else:
					inpackages.append(package)
			except KeyError,e:
				print e
				nopackages.append(package)
		if len(noupgradables)>0 or len(nopackages)>0:
			message = "<b>Some errors found:</b>\n"
			if len(noupgradables)>0:
				if len(noupgradables)>1:
					tpackages = ', '.join(noupgradables)
					message += "\nThe packages: '%s'\n <b>can't be upgraded</b> or the latest version is already installed"%tpackages
				else:
					message += "\nThe package: %s <b>can't be upgraded</b> or the latest version is already installed"%noupgradables[0]
			if len(nopackages)>0:
				if len(nopackages)>1:
					tpackages = ', '.join(nopackages)
					message += "\nThere are <b>no packages</b> called:\n '%s'"%tpackages
				else:
					message += '\nThere is <b>no package</b> called "%s"'%nopackages[0]
				dbuttons=gtk.BUTTONS_OK
			if len(inpackages)>0:
				if len(inpackages)>1:
					tpackages = ', '.join(inpackages)
					message += "\n<b>Only</b> these packages: '%s'\n will be installed"%tpackages

				else:
					message += "\n<b>Only</b>: %s will be installed"%inpackages[0]
				message +="\n\n<b>Continue?</b>"
				ans_exit = False
				dbuttons=gtk.BUTTONS_YES_NO
			else:
				message += "\n\nThere is <b>no package</b> to install"
				message += "\nInstallation will <b>not continue</b>"
				ans_exit = True
				dbuttons=gtk.BUTTONS_OK
			md = gtk.MessageDialog(parent=self,
			flags=gtk.DIALOG_MODAL | gtk.DIALOG_DESTROY_WITH_PARENT,
			type=gtk.MESSAGE_ERROR,
			buttons=dbuttons,
			message_format=None)
			md.set_markup(message)
			ans = md.run()
			print ans
			if ans == -9 or ans_exit:
				md.destroy()
				exit(0)
			md.destroy()
		print inpackages
		if len(inpackages)>0:
			for inpackage in inpackages:
				if cache[inpackage].is_upgradable:
					cache[inpackage].mark_upgrade()
				else:
					cache[inpackage].mark_install(auto_fix=True, auto_inst=True, from_user=True)
			try:
				self.set_size_request(800, 650)
				self.progress.show_terminal(expanded=True)
				cache.commit(self.progress.acquire, self.progress.install)					
			except LockFailedException,e:
				print e
				md = gtk.MessageDialog(parent=self,
				flags= gtk.DIALOG_MODAL | gtk.DIALOG_DESTROY_WITH_PARENT,
				type=gtk.MESSAGE_ERROR,
				buttons= gtk.BUTTONS_OK,
				message_format="You didn't run this program as root or another package\nmanager such as Synaptic or apt-get is running")
				md.run()
				md.destroy()
				self.progress.show_terminal(expanded=False)
				self.set_size_request(400, 200)
		else:
			md = gtk.MessageDialog(parent=self,
			flags=gtk.DIALOG_MODAL | gtk.DIALOG_DESTROY_WITH_PARENT,
			type=gtk.MESSAGE_ERROR,
			buttons=gtk.BUTTONS_OK,
			message_format="You didn't enter any package to install!")
			md.run()
			md.destroy()
		self.progress.show_terminal(expanded=False)
		self.set_size_request(400, 200)

if __name__ == '__main__':
	import sys
	if len(sys.argv)>2:
		packages = (' '.join(sys.argv[1:])).strip()
	elif len(sys.argv) >= 2:
		packages = sys.argv[1]
	else:
		packages = None
	ipd = InstallPackageDialog(packages)
	ipd.run()
